### ISSUE TITLE
nested impl Trait in trait with default definition does not compile

### ISSUE THREAD

**Issue by axos88** [rust-lang/rust#117104]
I tried this code:

```rust
trait Foo {
    fn foo() -> impl Stream<Item = impl Stream<Item = impl Future<Output = ()>>> {
        if true {
            unimplemented!()
        } else {
            let f = futures::future::ready(());
            let s1 = futures::stream::once(async { f });
            futures::stream::once(async { s1 })
        }
    }
}
```

I expected to see this happen: compiles, or a helpful error message

Instead, this happened: Confusing error message, there doesn't seem to be a future that could be awaited?...
Interestingly if the default implementation is removed, it compiles. Playground: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=6d9d947523afbecf604d1971f58bbc48
```
error[E0271]: type mismatch resolving `<impl Stream<Item = impl Future<Output = ()>> as Stream>::Item == impl Future<Output = ()>`
  --> /home/akos/projects/rust/tarpc/tarpc/src/server/incoming.rs:49:36
   |
49 |     fn foo() -> impl Stream<Item = impl Stream<Item = impl Future<Output = ()>>> {
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected associated type, found future
   |
   = help: consider `await`ing on both `Future`s
note: required by a bound in `Foo::{opaque#0}`
  --> /home/akos/projects/rust/tarpc/tarpc/src/server/incoming.rs:49:48
   |
49 |     fn foo() -> impl Stream<Item = impl Stream<Item = impl Future<Output = ()>>> {
   |                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Foo::{opaque#0}`

For more information about this error, try `rustc --explain E0271`.

```

### Meta

`rustc --version --verbose`:
```
rustc 1.75.0-nightly (1c05d50c8 2023-10-21)
binary: rustc
commit-hash: 1c05d50c8403c56d9a8b6fb871f15aaa26fb5d07
commit-date: 2023-10-21
host: x86_64-unknown-linux-gnu
release: 1.75.0-nightly
LLVM version: 17.0.3

```

<details><summary>Backtrace</summary>
<p>

```
N/A
```

</p>
</details>

**Comment by axos88** [comment]
The error message on the playground is a bit different, seems to be correlating with the actual problem more, but even more confusing:

```
   Compiling playground v0.0.1 (/playground)
error[E0271]: type mismatch resolving `<impl Stream<Item = impl Future<Output = ()>> as Stream>::Item == impl Future<Output = ()>`
  --> src/lib.rs:27:36
   |
27 |     fn foo() -> impl Stream<Item = impl Stream<Item = impl Future<Output = ()>>> {
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected associated type, found opaque type
   |
   = note: expected associated type `impl Future<Output = ()>` (trait associated opaque type at <src/lib.rs:27:55>)
                  found opaque type `impl Future<Output = ()>` (opaque type at <src/lib.rs:27:55>)
note: required by a bound in `Foo::{opaque#2}`
  --> src/lib.rs:27:48
   |
27 |     fn foo() -> impl Stream<Item = impl Stream<Item = impl Future<Output = ()>>> {
   |                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Foo::{opaque#2}`

For more information about this error, try `rustc --explain E0271`.
error: could not compile `playground` (lib) due to previous error
```

Expected and found are the same!

**Comment by compiler-errors** [comment]
Minimal:

```
use std::ops::Deref;

trait Foo {
    fn foo() -> impl Deref<Target = impl Deref<Target = impl Sized>> {
        &&()
    }
}

fn main() {}
```

Something funky going on with the default projection predicate installed for the GAT. I'll look into it 🤔

**Comment by axos88** [comment]
I'm impressed by the speedy handling of this issue! Was hoping this would get a fix in a few weeks . You're awesome!

**Comment by axos88** [comment]
Once the pr is merged, it will be available the next day on nightly, correct? Btw, the pr failed some tests.
